File: data-stuctures/arrays/find-the-duplicate-number.cs
public class Solution {
    public int FindDuplicate(int[] nums) {        
        HashSet<int> map = new HashSet<int>();
        
        for (int i = 0; i < nums.Length; i++)
        {
            if (map.Contains(nums[i]))
            {
                return nums[i];
            }
            
            map.Add(nums[i]);
        }
        
        return -1;
    }
}

File: data-stuctures/arrays/find_the_duplicate_number.py
'''https://leetcode.com/problems/find-the-duplicate-number/'''


def findDuplicate(self, nums: List[int]) -> int:
        compareSet = set()

        for i, val in enumerate(nums):

            if nums[i] not in compareSet:
                compareSet.add(nums[i])
            else:
                return nums[i]


def find_duplicates(nums):
    '''Solution using the Floyd’s Algorithm of cycle detection (Tortoise and Hare).'''

    tortoise = nums[0]
    hare = nums[0]

    while True:
        tortoise = nums[tortoise]
        hare = nums[nums[hare]]
        if tortoise == hare:
            break

    pointer = nums[0]

    while pointer != tortoise:
        pointer = nums[pointer]
        tortoise = nums[tortoise]

    return pointer